### ISSUE TITLE
loss saturate after several iteration

### ISSUE THREAD

**Issue by YuBeomGon** [HobbitLong/SupContrast#21]
Hi, thanks for your sharing.
I try to test your loss funtion, but is constant after some times.

the problem is that learning is going to wrong direction.
dot product of model's output is all, it means all vector is  grouped together regardless of label?

what should I check??
thanks.

-> loss
1.9459104537963867 
-> label
tensor([3, 5, 0, 1])       
-> dot production of augmented data's output
tensor([[1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000]],

1.9459099769592285
tensor([4, 4, 3, 3])
tensor([[1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000]],
       grad_fn=<MmBackward>)

**Comment by HobbitLong** [comment]
Are you using the CIFAR-10 or CIFAR-100 datasets that I tested?

**Comment by YuBeomGon** [comment]
No, I use text data, and bert model.
I fixed it by referring to your code.

I use agmentation also, 
so input dimension to loss ft is [batch, 2, z_dim]
I check the loss function also.

I dont know why all z output goes to one group regardless label.

**Comment by HobbitLong** [comment]
The `label` input needs to be integer vectors ranging from [0, num_class - 1].

**Comment by YuBeomGon** [comment]
yes,
my label is between 0 and 5, integer value (data has 6 class )
still I dont know the reason

**Comment by HobbitLong** [comment]
You could create some dummy data to debug. The current version of the loss looks correct to me.

**Comment by YuBeomGon** [comment]
ok.
if I check the reason, I will share here
thanks

**Comment by HobbitLong** [comment]
@YuBeomGon, thanks

**Comment by YuBeomGon** [comment]
I restart this issue, and finally solved.

I try to train text data using huggingface bert model.
in case of bert, augmentation is limited, so I didn't use augmentation and didnt use mlp for z embedding.
just use representation directly(cls token)
so input shape to supconloss is [batch size, 1, dim=768]

because of no augment data, one change is needed.
I add 1e-12 for blocking zero divide issue.
mean_log_prob_pos = (mask * log_prob).sum(1) / (mask.sum(1) + 1e-12) 

After lots of iteration, I check the loss steadily decreased.
it seemed that lots of iteration is needed because many label combination need to be tested.

thanks.

**Comment by HobbitLong** [comment]
so `mask.sum(1)` could be `0` in your case? And this is the issue that causes unstable training?

**Comment by YuBeomGon** [comment]
sorry for late reply.
yes, in my case, mask.sum(1) could be zero, because I didnt  use augmentation.
I just randomly sample, so  there is case which only one sample is per class.
